### ISSUE TITLE
strictProperties 

### ISSUE THREAD

**Issue by ghost** [ajv-validator/ajv#193]
it would be great to add strictProperties support
[https://github.com/fge/json-schema-validator/wiki/v5:-strictProperties](https://github.com/fge/json-schema-validator/wiki/v5:-strictProperties)

**Comment by epoberezkin** [comment]
I am not sure I understand what problem this keyword solves that is not already solved by existing keywords.

**Comment by ghost** [comment]
because additionnalProperties only apply to properties declared at the same level. 

In the following exemple, if my schema is a mix between 2 schema (part of each), i have no way currently to restrict to properties i require. The only way would be stictProperties.

``` json
{
    "$schema": "http://json-schema.org/schema#",
    "id": "http://xy.com/schema/schemaC.json",
    "allOf": [{
        "$ref": "schemaA.json"
    }, {
        "$ref": "schemaB.json"
    }, {
        "required": [
            "propA1",
            "propA3",
            "propA4",
            "propA7",
            "propB2",
            "propB4",
            "propB5"
        ],
        "strictProperties": true
    }]
}
```

With `"stricProperties" : true` i can not have a additionnal property not visible in `"required"`. If `"propA2"` or `"propC1"` is in my object, validation will fail.

**Comment by epoberezkin** [comment]
I see nothing in the proposal that would suggest it is the case.
I generally disagree with the idea of having keywords that have wider effects than the schema they are used in - there are quite a few edge cases that make it both unexpected and difficult to implement.

**Comment by epoberezkin** [comment]
According to the proposal, this schema:

``` json
{
  "required": [
    "propA1",
    "propA3",
    "propB2",
    "propB4",
  ],
  "strictProperties": true
}
```

is equivalent to:

``` json
{
  "properties": {
    "propA1": {},
    "propA3": {},
    "propB2": {},
    "propB4": {}
  },
  "required": [
    "propA1",
    "propA3",
    "propB2",
    "propB4",
  ],
  "additionalProperties": false
}
```

So it is just a syntax sugar that can be added with macro keyword definition. Can be added to [ajv-keywords](https://github.com/epoberezkin/ajv-keywords) I think

**Comment by ghost** [comment]
it is not equivalent.
"has no other properties than the ones defined in properties, patternPropertyGroups and required."

that means if i use a shema with external properties using $ref and i use required in the parent shema with "strictProperties" it will require and only accept properties in "required" property. With external properties, the property "additionalProperties" will not work as i would like.

**Comment by epoberezkin** [comment]
What do you mean by "external properties"? In any case strictProperties should be only affected by properties and other keywords in the same schema object, it can't take into account properties in other objects.

**Comment by epoberezkin** [comment]
Closing the issue. I will not be implementing any keywords with global visibility until they are added to the standard. And I am strongly against it, I think keywords should only have local visibility.

I haven't been presented with a single use-case for properties that cannot be expressed using existing JSON-schema vocabulary. I like the idea of [$merge/$patch](https://github.com/daveclayton/json-schema-validator/wiki/v5:-merge) keywords though, extending the schema is indeed not supported and it is needed to extend meta-schemas, for example.

If you are interested in strictProperties keyword with local visibility, we can add it to [ajv-keywords](https://github.com/epoberezkin/ajv-keywords) as I suggested (I already wrote the code, just need to add test cases...)
